Mathlib/Algebra/Ring/Basic.lean

```diff
@@ -174,20 +175,45 @@ lemma IsRightCancelMulZero.to_noZeroDivisors [MulZeroClass α]
   eq_zero_or_eq_zero_of_mul_eq_zero {_ y} h :=
     or_iff_not_imp_right.mpr fun ne ↦ mul_right_cancel₀ ne ((zero_mul y).symm ▸ h)
 
-instance (priority := 100) NoZeroDivisors.to_isCancelMulZero
-    [NonUnitalNonAssocRing α] [NoZeroDivisors α] :
-    IsCancelMulZero α where
-  mul_left_cancel_of_ne_zero ha _ _ h := by
-    rw [← sub_eq_zero, ← mul_sub] at h
-    exact sub_eq_zero.1 ((eq_zero_or_eq_zero_of_mul_eq_zero h).resolve_left ha)
-  mul_right_cancel_of_ne_zero hb _ _ h := by
-    rw [← sub_eq_zero, ← sub_mul] at h
-    exact sub_eq_zero.1 ((eq_zero_or_eq_zero_of_mul_eq_zero h).resolve_right hb)
+section NonUnitalNonAssocRing
+
+variable {R : Type*} [NonUnitalNonAssocRing R] {r : R}
+
+lemma isLeftRegular_iff_eq_zero_of_mul_left : IsLeftRegular r ↔ ∀ x, r * x = 0 → x = 0 where
+  mp h r' eq := h (by simp_rw [eq, mul_zero])
+  mpr h r₁ r₂ eq := sub_eq_zero.mp <| h _ <| by simp_rw [mul_sub, eq, sub_self]
+
+lemma isRightRegular_iff_eq_zero_of_mul_right : IsRightRegular r ↔ ∀ x, x * r = 0 → x = 0 where
+  mp h r' eq := h (by simp_rw [eq, zero_mul])
+  mpr h r₁ r₂ eq := sub_eq_zero.mp <| h _ <| by simp_rw [sub_mul, eq, sub_self]
+
+lemma isRegular_iff_eq_zero_of_mul :
+    IsRegular r ↔ (∀ x, r * x = 0 → x = 0) ∧ (∀ x, x * r = 0 → x = 0) := by
+  rw [isRegular_iff, isLeftRegular_iff_eq_zero_of_mul_left, isRightRegular_iff_eq_zero_of_mul_right]
+
+/-- A (not necessarily unital or associative) ring with no zero divisors has cancellative
+multiplicative on both sides. Since either left or right cancellative multiplication implies
+the absence of zero divisors, the four conditions are equivalent to each other. -/
+lemma noZeroDivisors_tfae : List.TFAE
+    [NoZeroDivisors R, IsLeftCancelMulZero R, IsRightCancelMulZero R, IsCancelMulZero R] := by
+  simp_rw [isLeftCancelMulZero_iff, isRightCancelMulZero_iff, isCancelMulZero_iff_forall_isRegular,
+    isLeftRegular_iff_eq_zero_of_mul_left, isRightRegular_iff_eq_zero_of_mul_right,
+    isRegular_iff_eq_zero_of_mul]
+  tfae_have 1 ↔ 2 := noZeroDivisors_iff_eq_zero_of_mul_left
+  tfae_have 1 ↔ 3 := noZeroDivisors_iff_eq_zero_of_mul_right
+  tfae_have 1 ↔ 4 := noZeroDivisors_iff_eq_zero_of_mul
+  tfae_finish
 
 /-- In a ring, `IsCancelMulZero` and `NoZeroDivisors` are equivalent. -/
-lemma isCancelMulZero_iff_noZeroDivisors [NonUnitalNonAssocRing α] :
-    IsCancelMulZero α ↔ NoZeroDivisors α :=
-  ⟨fun _ => IsRightCancelMulZero.to_noZeroDivisors _, fun _ => inferInstance⟩
+lemma isCancelMulZero_iff_noZeroDivisors : IsCancelMulZero R ↔ NoZeroDivisors R :=
```

Comment:
This already existed and came with a docstring.

Also I'm not sure why `IsLeft/RightCancelMulZero.to_noZeroDivisors` are not instances. If they are then there's no need of `instance (priority := 100) IsDomain.to_noZeroDivisors [Semiring α] [IsDomain α] : NoZeroDivisors α` below. I might try making them instances of priority 100 in another PR.